sycl/doc/Assert.md

```diff
@@ -0,0 +1,426 @@
+and runtime. The Low-Level Runtime is responsible for:
+ - detecting if assert failure took place;
+ - flushing assert message to `stderr` on host.
+
+When detected, Low-level Runtime reports assert failure to DPCPP Runtime
+via events objects. To achieve this, information about assert failure should be
+propagated from device-side to SYCL Runtime. This should be performed via calls
+to `piEventGetInfo`. This Plugin Interface call "lowers" to `clGetEventInfo` for
+OpenCL backend and `zeEventQueryStatus` for Level-Zero backend.
+
+Refer to [OpenCL](extensions/Assert/opencl.md) and [Level-Zero](extensions/Assert/level-zero.md)
+extensions.
+
+The following sequence of events describes how user code gets notified:
+ - Device side:
+   1. Assert fails in device-code in kernel
+      // It's not defined if GPU thread stops execution
+      // Other GPU threads are left untouched
+   2. Specialized version of `__devicelib_assert_fail` is called
+   3. Device immediately signals to host (Low-Level Runtime)
+ - Host side:
+   1. The assert failure gets detected by Low-Level Runtime
+   2. Low-Level Runtime prints assert failure message to `stderr`
+   3. Low-Level Runtime calls `abort()`
+
+
+### Fallback approach
+
+If Device-side Runtime doesn't support `__devicelib_assert_fail` then a buffer
+based approach comes in place. The approach doesn't require any support from
+Device-side Runtime and Native Device Compiler. Neither it does from Low-level
+Runtime.
+
+Within this approach, a dedicated assert buffer is allocated and implicit kernel
+argument is introduced. The argument is an accessor with `discard_read_write`
+or `discard_write` access mode. Accessor metadata is stored to program scope
+variable. This allows to refer to the accessor without modifying each and every
+user's function. Fallback implementation of `__devicelib_assert_fail` restores
+accessor metadata from program scope variable and writes assert information to
+the assert buffer. Atomic operations are used in order to not overwrite existing
+information.
+
+DPCPP Runtime checks contents of the assert buffer for assert failure flag after
+kernel finishes.
+
+The following sequence of events describes how user code gets notified:
+ - Device side:
+   1. Assert fails in device-code in kernel
+   2. Fallback version of `__devicelib_assert_fail` is called
+   3. Assert information is stored into assert buffer
+   4. Kernel continues running
+ - Host side:
+   1. A distinct thread is launched no later than the point of enqueue of the of
+      kernel with assertions
+   2. This thread polls the enqueued kernels for finish and checks the assert
+      buffer for assert data
+   3. If assert data is present DPCPP Runtime calls `abort()`
+
+
+#### Online-linking fallback `__devicelib_assert_fail`
+
+Online linking against fallback implementation of `__devicelib_assert_fail` is
+performed only when assertion is enabled.
```

Comment:
I suggest that in our design we rely on actual device capabilities (for which we already have al queries in place) instead of introducing `weak` linkage type usage, which is not supported by any backend/spec yet. I understand that there are some concerns about Level Zero API, but if it is intended to support online linking per its definition, then it seems clearer to me to explicitly check whether fallback is needed or not instead of relying on `weak` linkage type

Reply:
Clarified, that online-linking is only performed when device-side runtime doesn't support the required extension.